pkg/transport/timeout_listener.go

```diff
@@ -23,19 +23,32 @@ import (
 // If read/write on the accepted connection blocks longer than its time limit,
 // it will return timeout error.
 func NewTimeoutListener(addr string, scheme string, tlsinfo *TLSInfo, rdtimeoutd, wtimeoutd time.Duration) (net.Listener, error) {
-	ln, err := newListener(addr, scheme)
+	ln, err := newListener(addr, scheme, nil)
 	if err != nil {
 		return nil, err
 	}
-	ln = &rwTimeoutListener{
+	return newTimeoutListener(ln, scheme, rdtimeoutd, wtimeoutd, tlsinfo)
+}
+
+// NewTimeoutListerWithSocketOpts returns a listener that listens on the given address.
+// If read/write on the accepted connection blocks longer than its time limit,
+// it will return timeout error. Socket options can be passed and will be applied to the
+// ListenConfig.
+func NewTimeoutListerWithSocketOpts(addr string, scheme string, tlsinfo *TLSInfo, rdtimeoutd, wtimeoutd time.Duration, sopts *SocketOpts) (net.Listener, error) {
```

Comment:
This function parameters are getting too verbose. Can we make it variadic with something like `ListenerOption`? Then we can get rid of all those `tlsinfo` and others.

https://github.com/etcd-io/etcd/blob/3d7aac948b1cde9de96f24363c8a790883d56c1d/client/v3/op.go#L333-L339

Reply:
make sense thanks